### PR TITLE
feat: Add API key auth support on BeyondTrust provider

### PR DESCRIPTION
**Pull request by dtejadav**
## Problem Statement

What is the problem you're trying to solve?

This is a new feature to allow users to authenticate with Beyond Trust Provider either with API Key or Client Credentials.

## Related Issue

Fixes # BIPS-21956

## Proposed Changes

How do you like to solve the issue and why?
Adding the **ApiKey** attribute to BeyondtrustAuth and using it to authenticate when provided, otherwise the current authentication using Client Credentials would be used.

## Checklist

- [x] I have read the [contribution guidelines](https://external-secrets.io/latest/contributing/process/#submitting-a-pull-request)
- [x] All commits are signed with `git commit --signoff`
- [x] My changes have reasonable test coverage
- [x] All tests pass with `make test`
- [x] I ensured my PR is ready for review with `make reviewable`

### CODE REVIEW

**Comment by Skarlso** [comment]
@dtejadav Hello 👋 

Would you mind taking care of the complexity issue please? Thanks! :)

**Comment by dtejadav** [comment]
> @dtejadav Hello 👋
> 
> Would you mind taking care of the complexity issue please? Thanks! :)

Hello @Skarlso 👋, thank you, complexity issue has been solved, could you check again the PR?

File: apis/externalsecrets/v1beta1/secretstore_beyondtrust_types.go [hunk 1]
```diff
@@ -29,11 +29,13 @@ type BeyondTrustProviderSecretRef struct {
 
 // Configures a store to sync secrets using BeyondTrust Password Safe.
 type BeyondtrustAuth struct {
-	// +required - API OAuth Client ID.
-	ClientID *BeyondTrustProviderSecretRef `json:"clientId"`
-	// +required - API OAuth Client Secret.
-	ClientSecret *BeyondTrustProviderSecretRef `json:"clientSecret"`
-	// Content of the certificate (cert.pem) for use when authenticating with an OAuth client Id using a Client Certificate.
+	// API Key. If API key is not provided then ClientID/ClientSecret become required.
```
Comment (Skarlso) [inline apis/externalsecrets/v1beta1/secretstore_beyondtrust_types.go:32]:
```suggestion
	// APIKey If not provided then ClientID/ClientSecret become required.
```

File: apis/externalsecrets/v1beta1/secretstore_beyondtrust_types.go [hunk 2]
```diff
@@ -29,11 +29,13 @@ type BeyondTrustProviderSecretRef struct {
...
+	APIKey *BeyondTrustProviderSecretRef `json:"apiKey,omitempty"`
+	// API OAuth Client ID.
```
Comment (Skarlso) [inline apis/externalsecrets/v1beta1/secretstore_beyondtrust_types.go:34]:
```suggestion
	// ClientID is the API OAuth Client ID.
```

File: apis/externalsecrets/v1beta1/secretstore_beyondtrust_types.go [hunk 3]
```diff
@@ -29,11 +29,13 @@ type BeyondTrustProviderSecretRef struct {
...
+	ClientID *BeyondTrustProviderSecretRef `json:"clientId,omitempty"`
+	// API OAuth Client Secret.
```
Comment (Skarlso) [inline apis/externalsecrets/v1beta1/secretstore_beyondtrust_types.go:36]:
```suggestion
	// ClientSecret is the API OAuth Client Secret.
```

File: apis/externalsecrets/v1beta1/secretstore_beyondtrust_types.go [hunk 4]
```diff
@@ -29,11 +29,13 @@ type BeyondTrustProviderSecretRef struct {
...
+	ClientSecret *BeyondTrustProviderSecretRef `json:"clientSecret,omitempty"`
+	// Content o f the certificate (cert.pem) for use when authenticating with an OAuth client Id using a Client Certificate.
```
Comment (Skarlso) [inline apis/externalsecrets/v1beta1/secretstore_beyondtrust_types.go:38]:
```suggestion
	// Certificate (cert.pem) for use when authenticating with an OAuth client Id using a Client Certificate.
```

File: pkg/provider/beyondtrust/provider.go [hunk 5]
```diff
@@ -197,6 +184,79 @@ func (p *Provider) NewClient(ctx context.Context, store esv1beta1.GenericStore,
...
+func validateInputs(params utils.ValidationParams) error {
+	return utils.ValidateInputs(params)
+}
+
+func getAuthenticator(input AuthenticatorInput) (*auth.AuthenticationObj, error) {
+    if input.Config.Auth.APIKey != nil {
+        return auth.AuthenticateUsingApiKey(input.HttpClientObj, input.BackoffDefinition, input.ApiURL, input.Logger, input.RetryMaxElapsedTimeMinutes, input.ApiKey)
+    } else {
+        return auth.Authenticate(input.HttpClientObj, input.BackoffDefinition, input.ApiURL, input.ClientID, input.ClientSecret, input.Logger, input.RetryMaxElapsedTimeMinutes)
+    }
```
Comment (Skarlso) [inline pkg/provider/beyondtrust/provider.go:257]:
```suggestion
    if input.Config.Auth.APIKey != nil {
        return auth.AuthenticateUsingApiKey(input.HttpClientObj, input.BackoffDefinition, input.ApiURL, input.Logger, input.RetryMaxElapsedTimeMinutes, input.ApiKey)
    }

    return auth.Authenticate(input.HttpClientObj, input.BackoffDefinition, input.ApiURL, input.ClientID, input.ClientSecret, input.Logger, input.RetryMaxElapsedTimeMinutes)
```

**Comment by Skarlso** [comment]
Hm, something's not quite right with the rebase... :)

**Comment by dtejadav** [comment]
Working on these linter errors: https://github.com/external-secrets/external-secrets/actions/runs/11918402553/job/33216221459

**Comment by dtejadav** [comment]
@Skarlso, linting errors were solved, could you please check the PR again?

**Review by Skarlso — APPROVED** [review]

**Comment by Skarlso** [comment]
@dtejadav Could you please try to fix the rebase history? It feels super weird to squash that into main. :D

**Comment by dtejadav** [comment]
> @dtejadav Could you please try to fix the rebase history? It feels super weird to squash that into main. :D

Hi @Skarlso, I fixed rebase history, hope it's simpler now, could you check again please? 🔍

**Comment by Skarlso** [comment]
Sadly no man. :( I'm so sorry. Looks at your change set.. It's now everything. :/ 

Try doing a git reset --hard to your previous hash and then push that with force.

Then save your diff and open another PR from main. That's the cleanest man.

**Review by Skarlso — APPROVED** [review]

**Pull request merged**